### PR TITLE
json-functions: Update status and links

### PR DESCRIPTION
**Pull request by dveeden**
### What is changed, added or deleted? (Required)

Update the status of JSON functions in TiDB

```
mysql> CREATE TABLE t3 (id int primary key auto_increment, o_id int, attribute varchar(255), value varchar(255));
Query OK, 0 rows affected (0.12 sec)

mysql> INSERT INTO t3(o_id, attribute, value) VALUES(2,'color','red'),(2,'fabric','silk'),(3,'color','green'),(3,'shape','square');
Query OK, 4 rows affected (0.02 sec)
Records: 4  Duplicates: 0  Warnings: 0

mysql> SELECT o_id, attribute, value FROM t3;
+------+-----------+--------+
| o_id | attribute | value  |
+------+-----------+--------+
|    2 | color     | red    |
|    2 | fabric    | silk   |
|    3 | color     | green  |
|    3 | shape     | square |
+------+-----------+--------+
4 rows in set (0.00 sec)

mysql> SELECT o_id, JSON_ARRAYAGG(attribute) AS attributes
    -> FROM t3 GROUP BY o_id;
+------+---------------------+
| o_id | attributes          |
+------+---------------------+
|    3 | ["color", "shape"]  |
|    2 | ["color", "fabric"] |
+------+---------------------+
2 rows in set (0.01 sec)

mysql> SELECT tidb_version()\G
*************************** 1. row ***************************
tidb_version(): Release Version: v6.0.0
Edition: Community
Git Commit Hash: 36a9810441ca0e496cbd22064af274b3be771081
Git Branch: heads/refs/tags/v6.0.0
UTC Build Time: 2022-03-31 10:33:28
GoVersion: go1.18
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false
1 row in set (0.01 sec)

mysql> SELECT JSON_PRETTY('{"k":"v"}');
+--------------------------+
| JSON_PRETTY('{"k":"v"}') |
+--------------------------+
| {
  "k": "v"
}           |
+--------------------------+
1 row in set (0.00 sec)

mysql> SELECT JSON_MERGE_PATCH('[1, 2]', '[true, false]');
+---------------------------------------------+
| JSON_MERGE_PATCH('[1, 2]', '[true, false]') |
+---------------------------------------------+
| [true, false]                               |
+---------------------------------------------+
1 row in set (0.00 sec)
```

### Which TiDB version(s) do your changes apply to? (Required)

**Tips for choosing the affected version(s):**

By default, **CHOOSE MASTER ONLY** so your changes will be applied to the next TiDB major or minor releases. If your PR involves a product feature behavior change or a compatibility change, **CHOOSE THE AFFECTED RELEASE BRANCH(ES) AND MASTER**.

For details, see [tips for choosing the affected versions](https://github.com/pingcap/docs/blob/master/CONTRIBUTING.md#guideline-for-choosing-the-affected-versions).

- [x] master (the latest development version)
- [ ] v6.1 (TiDB 6.1 versions)
- [ ] v6.0 (TiDB 6.0 versions)
- [ ] v5.4 (TiDB 5.4 versions)
- [ ] v5.3 (TiDB 5.3 versions)
- [ ] v5.2 (TiDB 5.2 versions)
- [ ] v5.1 (TiDB 5.1 versions)
- [ ] v5.0 (TiDB 5.0 versions)
- [ ] v4.0 (TiDB 4.0 versions)

### What is the related PR or file link(s)?

- This PR is translated from:
- Other reference link(s):

### Do your changes match any of the following descriptions?

- [ ] Delete files
- [ ] Change aliases
- [ ] Need modification after applied to another branch 
- [ ] Might cause conflicts after applied to another branch

### CODE REVIEW

**Comment by dveeden** [comment]
Note that https://github.com/pingcap/tidb/issues/7968 lists this as not-yet-implemented JSON support:

- [JSON_TABLE](https://mysqlserverteam.com/json_table-the-best-of-both-worlds/) function + add https://github.com/pingcap/tidb/issues/7546 that were backported to 5.7
- multi-valued indexes on json arrays / json_overlaps() and member of()
- JSON schema validation
- JSON_VALUE function (8.0.21)

File: functions-and-operators/json-functions.md [hunk 1]
```diff
@@ -103,6 +98,8 @@ The following JSON functions are unsupported in TiDB. You can track the progress
```
Comment (mjonss) [inline functions-and-operators/json-functions.md:98]: I don't think this renders correctly, please format this whole paragraph as the two first bullet points.

**Review by mjonss — APPROVED** [review]
LGTM, with some suggestion on fixing rendering issues.

**Comment by mjonss** [comment]
json_pretty included in [#24675](https://github.com/pingcap/tidb/pull/24675).
json_merge_patch included in [#24711](https://github.com/pingcap/tidb/pull/24711).
json_arrayagg included in [#19957](https://github.com/pingcap/tidb/pull/19957).

**Review by tangenta — APPROVED** [review]

File: functions-and-operators/json-functions.md (hunk 1 above)
Comment (dveeden) [reply functions-and-operators/json-functions.md:98]: I think the context of your comment got lost somehow. It also looks like it is already fixed. Could you check if this is really the case and resolve this discussion?

**Review by mjonss — APPROVED** [review]
Suggestion to fix the formatting/rendering of 'See also' paragraph.

**Comment by dveeden** [comment]
> Suggestion to fix the formatting/rendering of 'See also' paragraph.

No this is as intended.

The lines like this should not end up in the doc directly:
```markdown
[json_pretty]: https://dev.mysql.com/doc/refman/5.7/en/json-utility-functions.html#function_json-pretty
```

But this allows lines like this to reference the URL:
```markdown
| [JSON_PRETTY(json_doc)][json_pretty] | Pretty formatting of a JSON document |
```

Another option would be to combine the two lines like this:
```markdown
| [JSON_PRETTY(json_doc)][https://dev.mysql.com/doc/refman/5.7/en/json-utility-functions.html#function_json-pretty] | Pretty formatting of a JSON document |
```

But then the markdown table gets difficult to read due to the long lines.

The problem with formatting that I fixed was related to this, there were functions that had a reference that was not defined, so it couldn't format the link properly.

**Comment by dveeden** [comment]
PTAL @TomShawn @qiancai

**Comment by dveeden** [comment]
/cherry-pick release-6.1
/cherry-pick release-5.4
/cherry-pick release-5.3
/cherry-pick release-5.2
/cherry-pick release-5.1

v6.1.0
```
sql> SELECT JSON_MERGE_PATCH('[1, 2]', '[true, false]') a, JSON_PRETTY('{}') b, (SELECT JSON_ARRAYAGG(x) FROM (SELECT 'a' x UNION ALL SELECT 'b') a) c, tidb_version()\G
*************************** 1. row ***************************
             a: [true, false]
             b: {}
             c: ["a", "b"]
tidb_version(): Release Version: v6.1.0
Edition: Community
Git Commit Hash: 1a89decdb192cbdce6a7b0020d71128bc964d30f
Git Branch: heads/refs/tags/v6.1.0
UTC Build Time: 2022-06-05 05:15:11
GoVersion: go1.18.2
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false
1 row in set (0.0023 sec)
```

v6.0.0
```
sql> SELECT JSON_MERGE_PATCH('[1, 2]', '[true, false]') a, JSON_PRETTY('{}') b, (SELECT JSON_ARRAYAGG(x) FROM (SELECT 'a' x UNION ALL SELECT 'b') a) c, tidb_version()\G
*************************** 1. row ***************************
             a: [true, false]
             b: {}
             c: ["a", "b"]
tidb_version(): Release Version: v6.0.0
Edition: Community
Git Commit Hash: 36a9810441ca0e496cbd22064af274b3be771081
Git Branch: heads/refs/tags/v6.0.0
UTC Build Time: 2022-03-31 10:33:28
GoVersion: go1.18
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false
1 row in set (0.0028 sec)
```

v5.4.1
```
sql> SELECT JSON_MERGE_PATCH('[1, 2]', '[true, false]') a, JSON_PRETTY('{}') b, (SELECT JSON_ARRAYAGG(x) FROM (SELECT 'a' x UNION ALL SELECT 'b') a) c, tidb_version()\G
*************************** 1. row ***************************
             a: [true, false]
             b: {}
             c: ["a", "b"]
tidb_version(): Release Version: v5.4.1
Edition: Community
Git Commit Hash: be484475e76b3aaa2cad520111353c3327db35fa
Git Branch: heads/refs/tags/v5.4.1
UTC Build Time: 2022-05-07 04:31:32
GoVersion: go1.16.4
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false
1 row in set (0.0024 sec)
```

v5.3.1
```
sql> SELECT JSON_MERGE_PATCH('[1, 2]', '[true, false]') a, JSON_PRETTY('{}') b, (SELECT JSON_ARRAYAGG(x) FROM (SELECT 'a' x UNION ALL SELECT 'b') a) c, tidb_version()\G
*************************** 1. row ***************************
             a: [true, false]
             b: {}
             c: ["a", "b"]
tidb_version(): Release Version: v5.3.1
Edition: Community
Git Commit Hash: 459917c6f83df155edcc03f6ebde24942ff73b0e
Git Branch: heads/refs/tags/v5.3.1
UTC Build Time: 2022-03-02 08:35:22
GoVersion: go1.16.4
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false
1 row in set (0.0022 sec)
```

v5.2.4
```
sql> SELECT JSON_MERGE_PATCH('[1, 2]', '[true, false]') a, JSON_PRETTY('{}') b, (SELECT JSON_ARRAYAGG(x) FROM (SELECT 'a' x UNION ALL SELECT 'b') a) c, tidb_version()\G
*************************** 1. row ***************************
             a: [true, false]
             b: {}
             c: ["a", "b"]
tidb_version(): Release Version: v5.2.4
Edition: Community
Git Commit Hash: fe5aa9760b587799361c954cffab4d786dbdbf7e
Git Branch: heads/refs/tags/v5.2.4
UTC Build Time: 2022-04-18 06:22:05
GoVersion: go1.16.4
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false
1 row in set (0.0023 sec)
```

v5.1.3
```
sql> SELECT JSON_MERGE_PATCH('[1, 2]', '[true, false]') a, JSON_PRETTY('{}') b, (SELECT JSON_ARRAYAGG(x) FROM (SELECT 'a' x UNION ALL SELECT 'b') a) c, tidb_version()\G
ERROR: 1305 (42000): FUNCTION JSON_MERGE_PATCH does not exist

sql> SELECT JSON_PRETTY('{}') b, (SELECT JSON_ARRAYAGG(x) FROM (SELECT 'a' x UNION ALL SELECT 'b') a) c, tidb_version()\G
ERROR: 1105 (HY000): unsupported agg function: json_arrayagg

sql> SELECT JSON_PRETTY('{}') b, tidb_version()\G
*************************** 1. row ***************************
             b: {}
tidb_version(): Release Version: v5.1.3
Edition: Community
Git Commit Hash: dc079337ef0c9d78eec2d2acb1eb31e12117d29b
Git Branch: heads/refs/tags/v5.1.3
UTC Build Time: 2021-11-26 01:30:03
GoVersion: go1.16.4
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false
1 row in set (0.0012 sec)
```

v5.0.6
```
sql> SELECT JSON_PRETTY('{}') b, tidb_version()\G
ERROR: 1305 (42000): FUNCTION JSON_PRETTY does not exist

sql> SELECT tidb_version()\G
*************************** 1. row ***************************
tidb_version(): Release Version: v5.0.6
Edition: Community
Git Commit Hash: 6416f8d601472892d245b950dfd5547e857a1a33
Git Branch: heads/refs/tags/v5.0.6
UTC Build Time: 2021-12-23 12:26:47
GoVersion: go1.13
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false
1 row in set (0.0009 sec)
```

**Comment by dveeden** [comment]
I tried a simple script to detect if there are more issues like this in other pages.

```
[dvaneeden@dve-carbon docs]$ git checkout upstream/release-5.0
...
[dvaneeden@dve-carbon docs]$ awk -f /tmp/links.awk functions-and-operators/json-functions.md
No target defined for "json_objectagg" Link defined as: | [JSON_OBJECTAGG(key, value)][json_objectagg] | Provides an aggregation of values for a given key. |
No target defined for "json_storage_size" Link defined as: | [JSON_STORAGE_SIZE(json_doc)][json_storage_size] | Returns an approximate size of bytes required to store the json value. As the size does not account for TiKV using compression, the output of this function is not strictly compatible with MySQL. |
```

However it doesn't look like other pages have the same issue and there are some false positives, so this is not something we can easily add to CI/CD.

```awk
/\[.*\]\[.*\]/ {
	a=substr($0,index($0, "][")+2);
	links[substr(a,1,index(a,"]")-1)] = $0
}
/^\[.*\]: / {
	targets[substr($0,2,index($0, "]")-2)] = $0
}

END {
	for (link in links) {
		if (targets[link] == "") {
			print "No target defined for \"" link "\"", "Link defined as: " links[link]
		}
	}
}
```

**Pull request merged**